### PR TITLE
90% Plat 1729 refactor for testing

### PR DESCRIPTION
**Pull request by ShaunParsons**
This PR refactors obtaining new tokens.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,19 +25,21 @@
     "coverage-html": "nyc --reporter=html npm run test",
```
Comment (ShaunParsons) [inline package.json]: Need a version bump

File: test/unit/public-key-store.js [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+                    generateTokenStub.calledWith(personaClient, testObtainTokenCacheKey, testXRequestId, testOpts.id, testOpts.secret).should.equal(true);
+
+                    publicKey.access_token.should.equal(testPublicKeyObject.access_token);
+                    publicKey.expires_at.should.equal(testUnexpiredTimestamp);
+                    done();
+                });
+            });
+            
+            // https://github.com/talis/platform/issues/1729
+            it('cache error', function(done) {
```
Comment (ShaunParsons) [inline test/unit/public-key-store.js]: This is the test for the functional change in this PR. If the redis cache has an error, fallback to getting the token from Persona.

File: client/lib/public-key-store.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    var secret = opts.secret;
+    var xRequestId = opts.xRequestId || uuid.v4();
+
+    var cacheKey = client._formatCacheKey(
+        "obtain_token:" + hashKey(id)
+    );
+
+    // try cache first
+    client.tokenCache.get(cacheKey, function (err, reply) {
+        if (reply) {
```
Comment (ShaunParsons) [inline client/lib/public-key-store.js]: This is the functional change. Previously we had:

```js
    // try cache first
    this.tokenCache.get(cacheKey, function (err, reply) {
        if (err) {
            callback(err, null);
        } else {
if (reply == null) {```

So now, unless the reply is valid we generate a new token from persona.

File: test/unit/hash-key.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+"use strict";
```
Comment (ShaunParsons) [inline test/unit/hash-key.js]: These tests probably aren't needed... I can remove if you want.

File: client/lib/generate-token.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
+var querystring = require('querystring');
+
+var generateToken = function (client, cacheKey, xRequestId, id, secret, callback) {
```
Comment (markwallsgrove) [inline client/lib/generate-token.js]:
```suggestion
function generateToken(client, cacheKey, xRequestId, id, secret, callback) {
```

Comment (markwallsgrove) [reply client/lib/generate-token.js]: to name the function

Comment (markwallsgrove) [reply client/lib/generate-token.js]: for debug info

File: client/lib/public-key-store.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
+var _ = require('lodash');
+var uuid = require('uuid');
+var querystring = require("querystring");
+
+var hashKey = require('./hash-key');
+var getRemotePublicKey = require('./remote-public-key');
+var generateToken = require('./generate-token');
+
+var setCachedPublicKey = function setCachedPublicKey(client, cacheKey, publicKey) {
```
Comment (markwallsgrove) [inline client/lib/public-key-store.js]:
```suggestion
function setCachedPublicKey(client, cacheKey, publicKey) {
```

File: client/lib/public-key-store.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    client.log('debug', 'Caching public key for ' + client.config.cert_timeout_sec + 's');
+    client.tokenCache.set(cacheKey, publicKey, client.config.cert_timeout_sec);
+}
+
+var getCachedPublicKey = function getCachedPublicKey(client, cacheKey, refresh, xRequestId, callback) {
```
Comment (markwallsgrove) [inline client/lib/public-key-store.js]:
```suggestion
function getCachedPublicKey(client, cacheKey, refresh, xRequestId, callback) {
```

File: client/lib/public-key-store.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                }
+                setCachedPublicKey(client, cacheKey, publicKey);
+                callback(null, publicKey);
+                return;
+            });
+        });
+    }
+}
+
+var obtainToken = function obtainToken(client, opts, callback) {
```
Comment (markwallsgrove) [inline client/lib/public-key-store.js]:
```suggestion
function obtainToken(client, opts, callback) {
```

File: client/lib/remote-public-key.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+var uuid = require('uuid');
+var ERROR_TYPES = require('./error-types');
+
+var getRemotePublicKey = function getRemotePublicKey(client, cacheKey, xRequestId, callback) {
```
Comment (markwallsgrove) [inline client/lib/remote-public-key.js]:
```suggestion
function getRemotePublicKey(client, cacheKey, xRequestId, callback) {
```

File: client/index.js [hunk 10]
```diff
@@ -626,122 +558,7 @@ PersonaClient.prototype.isPresignedUrlValid = function (presignedUrl, secret) {
...
-                            id: id,
-                            secret: secret,
-                            xRequestId: xRequestId
-                        }, callback);
-                    }
-                });
-            }
-        }
-    }.bind(this));
+    publicKeyStore.obtainToken(this, opts, callback);
```
Comment (markwallsgrove) [inline client/index.js]: Not sure what I think of handing in `this` as it tightly bounds the client to the methods being called, it feels a bit odd.

Comment (ShaunParsons) [reply client/index.js]: Its actually a fairly common way of doing things in nodes. It makes functions like `obtainToken` really easy to unit test by stubbing out the client.

Comment (markwallsgrove) [reply client/index.js]: I guess what I dislike is not knowing what the client should look like.

**Pull request merged**